gems/smithy/lib/smithy/views/client/plugin_list.rb

```diff
@@ -0,0 +1,84 @@
+# frozen_string_literal: true
+
+module Smithy
+  module Views
+    module Client
+      # @api private
+      class PluginList
+        include Enumerable
+
+        def initialize(plan)
+          @plan = plan
+          @plugins = default_plugins + transport_plugins('http') + generated_plugins
+          @plugins.each do |plugin|
+            require_path = plugin.require_path
+            next unless require_path
+
+            require_path = File.absolute_path(require_path) unless plugin.relative_path?
+            Kernel.require(require_path)
+          end
+        end
+
+        def each(&)
+          @plugins.each(&)
+        end
+
+        private
+
+        def namespace
+          Util::Namespace.namespace_from_gem_name(@plan.options[:gem_name])
+        end
+
+        def gem_name
+          @plan.options[:gem_name]
+        end
+
+        def gem_dir
+          @plan.options[:destination_root]
+        end
+
+        def default_plugins
+          Smithy::Client::Base.plugins.map do |plugin|
+            Plugin.new(class_name: plugin.name, require_path: nil, default: true)
+          end
+        end
+
+        # def weld_plugins
+        #   plugins = @plan.welds.map(&:plugins).reduce({}, :merge)
+        #   plugins.map do |class_name, path|
+        #     Plugin.new(class_name: class_name, path: path)
+        #   end
+        # end
+
+        def generated_plugins
+          # Nested namespaces for Plugins will not load unless we define them.
+          define_namespaces
+          plugins = {
+            "#{namespace}::Plugins::Endpoint" => "#{gem_dir}/lib/#{gem_name}/plugins/endpoint.rb"
+          }
+          plugins.map do |class_name, require_path|
+            Plugin.new(class_name: class_name, require_path: require_path)
+          end
+        end
+
+        def transport_plugins(protocol)
+          plugins = {
+            'http' => { 'Smithy::Client::Plugins::NetHTTP' => 'smithy-client/plugins/net_http' }
```

Comment:
This feels like it should be a constant somewhere.... and long term should maybe be extensible by welds.  But I think fine to leave as a constant for now.

Reply:
Transport would be configured somehow and index into this. I think protocol would work the same way. I agree it's fine for now..